controllers/orders.js

```diff
@@ -0,0 +1,166 @@
+const psUpdateFrequents = new PS({ name: 'update-freq', text: `UPDATE Frequents set lastUsed = CURRENT_TIMESTAMP WHERE uid = $1 AND addrId = $2` })
+const psAddAddress = new PS({
+    name: 'add-address',
+    text: `INSERT INTO Address(streetName, unit, postalCode)
+            SELECT $1, $2, $3
+            EXCEPT
+            SELECT streetname, unit, postalCode FROM Address
+            RETURNING addrId`})
+const psAddOrder = new PS({
+    name: 'add-order',
+    text: `Insert INTO Orders(customerId, orderTime, deliveryFee, isDeliveryFeeWaived, finalPrice, addrId, pid) VALUES
+            ($1, CURRENT_TIMESTAMP, $2, $3, $4, $5, $6) RETURNING oid` })
+const psAddPoints = new PS({
+    name: 'add-points',
+    text: `UPDATE Customers set points = (points + $2) WHERE uid = $1`
+})
+const psAddCollates = new PS({
+    name: 'add-collates',
+    text: `INSERT INTO Collates(fname, rid, oid, totalPrice, qty)
+            SELECT $1 fname, $2 rid, $3 oid, T.totalPrice, $4 qty
+            FROM (SELECT (price * $4) totalPrice
+                  FROM Food WHERE rid = $2 AND fname = $1) T`
+})
+const psMinusWaivePoints = new PS({
+    name: 'minus-waive-points',
+    text: `UPDATE Customers set points = (points - $2) WHERE uid = $1 AND points >= $2`
+})
+const psUpdateFoodCount = new PS({
+    name: 'update-food-count',
+    text: `UPDATE Food set numOrders = (numOrders + $3) WHERE fname = $1 and rid = $2`
+})
+
+const addOrder = async function (user, order) {
+    let total = await totalPrice(order.rid, order.items);
+    let addrId = null;
+    let awardedPoints = 0;
+    if (!(await validateOrder(user, order, total))) {
+        throw "Invalid Order";
+    }
+    const q1 = await db.oneOrNone(psAddAddress,
+        [order.addrInfo.address, order.addrInfo.unit, order.addrInfo.postal]);
+    if (q1 != null) {
+        addrId = q1.addrid;
+    } else {
+        const q = await db.oneOrNone(psGetAddrId,
+            [order.addrInfo.address, order.addrInfo.unit, order.addrInfo.postal]);
+        addrId = q.addrid;
+    }
+    if (order.pid != null) {
+        const { discountedPrice, points } = await getDiscountedPriceAndPoints(total, order.pid)
+        awardedPoints = points;
+        total = discountedPrice;
+    }
+    await processOrder(user, order, addrId, total, awardedPoints);
+}
+
+async function processOrder(uid, order, addrId, total, awardedPoints) {
+    const isFreq = await isFrequents(uid, addrId);
+    try {
+        await db.tx(async t => {
+            const queries = [];
+            if (isFreq) {
+                queries.push(t.none(psUpdateFrequents, [uid, addrId]));
+            } else {
+                queries.push(t.none(psAddFrequents, [uid, addrId]));
+            }
+            //Add order
+            const qAddOrder = await t.one(psAddOrder, [uid, 3, order.waiveFee, total, addrId, order.pid])
+            let oid = qAddOrder.oid;
+            queries.push(qAddOrder);
```

Comment:
Is it necessary to push it in if it's already been awaited?

Reply:
I don't think so